### PR TITLE
Bluetooth: Controller: Introduce scan_aux_chain context

### PR DESCRIPTION
**Pull request by Tronil**
Introduces the concept of ll_scan_aux_chain, that allows for keeping track of several advertising chains with a minimal overhead. This replaces using several scan aux sets each with an associated ticker IDs

Only one ticker is used for following chains; This is achieved by having the scheduled chains in a sorted list and restarting the ticker when the first in the list expires

EVENT_DONE_EXTRA_TYPE_SCAN_AUX now includes a pointer to the lll structure to identify which chain to flush (to avoid race conditions)

New implementation is hidden behind a config for now (defaults
to off): BT_CTLR_SCAN_AUX_USE_CHAINS

Other minor changes:
- rx_incomplete node pointer was moved to ll_sync_set
- data_len removed from ll_sync_set (it uses the chains instead)

### CODE REVIEW

**Comment by cvinayak** [comment]
This is very welcome improvement to the Controller, hope my recently PRs are not too much of conflict?

Is it possible to keep old design conditionally compiled out for two zephyr releases, i.e. old design remains as default for one release cycle (Zephyr v4.0.99) and new design for the next (Zephyr v4.1.99), and we delete the old one subsequently (Zephyr v4.2.0). We have design listing with the current design and hence, this will allow users to compare the performance improvements, and us to have some time to ensure there is not regression in conformance.

**Comment by Tronil** [comment]
> This is very welcome improvement to the Controller, hope my recently PRs are not too much of conflict?

It should be fine I think - I just need to rebase once more.

> Is it possible to keep old design conditionally compiled out for two zephyr releases, i.e. old design remains as default for one release cycle (Zephyr v4.0.99) and new design for the next (Zephyr v4.1.99), and we delete the old one subsequently (Zephyr v4.2.0). We have design listing with the current design and hence, this will allow users to compare the performance improvements, and us to have some time to ensure there is no regression in conformance.

Sure that should be doable; Though in that case I think it will be better to basically switch out the entire scan_aux.c implementation - otherwise it will get very messy. I'll get that done.

**Comment by cvinayak** [comment]
@jakkra This PR should give higher extended adv report rate with lower RAM requirements, compare to use of BT_CTLR_SCAN_AUX_SET.

**Comment by Tronil** [comment]
New implementation is now hidden behind `config BT_CTLR_SCAN_AUX_USE_CHAINS` (defaulting to off). We should have at least a few tests enabling it though; I'll look at that next.

**Comment by cvinayak** [comment]
Put it ready for review so the clock gets ticking towards merge-able state for v4.0.0.

**Review by erbr-ot — DISMISSED** [review]
Solid!

**Comment by cvinayak** [comment]
Comparing support for 9 simultaneous extended advertising reception support:

```
[181/182] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       68736 B         1 MB      6.56%
             RAM:       28812 B       256 KB     10.99%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/jack/workspace/zephyrproject/zephyr/build/observer_nrf52840/zephyr/zephyr.elf for board: nrf52840dk
(python) [jack@vich-archlinux-dell-laptop observer_nrf52840]$ git diff
diff --git a/samples/bluetooth/observer/prj_extended.conf b/samples/bluetooth/observer/prj_extended.conf
index cab86c38d14..fed3afef1c7 100644
--- a/samples/bluetooth/observer/prj_extended.conf
+++ b/samples/bluetooth/observer/prj_extended.conf
@@ -14,4 +14,9 @@ CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX=1650
 
 # Increase Zephyr Bluetooth LE Controller Rx buffer to receive complete chain
 # of PDUs
-CONFIG_BT_CTLR_RX_BUFFERS=9
+CONFIG_BT_CTLR_RX_BUFFERS=18
+
+CONFIG_BT_CTLR_ADVANCED_FEATURES=y
+CONFIG_BT_CTLR_SCAN_AUX_SET=9
+# CONFIG_BT_CTLR_SCAN_AUX_USE_CHAINS=y
+# CONFIG_BT_CTLR_SCAN_AUX_CHAIN_COUNT=9
```

```
[181/182] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       69412 B         1 MB      6.62%
             RAM:       28364 B       256 KB     10.82%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/jack/workspace/zephyrproject/zephyr/build/observer_nrf52840/zephyr/zephyr.elf for board: nrf52840dk
(python) [jack@vich-archlinux-dell-laptop observer_nrf52840]$ git diff
diff --git a/samples/bluetooth/observer/prj_extended.conf b/samples/bluetooth/observer/prj_extended.conf
index cab86c38d14..7dbe1c22f46 100644
--- a/samples/bluetooth/observer/prj_extended.conf
+++ b/samples/bluetooth/observer/prj_extended.conf
@@ -14,4 +14,9 @@ CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX=1650
 
 # Increase Zephyr Bluetooth LE Controller Rx buffer to receive complete chain
 # of PDUs
-CONFIG_BT_CTLR_RX_BUFFERS=9
+CONFIG_BT_CTLR_RX_BUFFERS=18
+
+CONFIG_BT_CTLR_ADVANCED_FEATURES=y
+# CONFIG_BT_CTLR_SCAN_AUX_SET=9
+CONFIG_BT_CTLR_SCAN_AUX_USE_CHAINS=y
+CONFIG_BT_CTLR_SCAN_AUX_CHAIN_COUNT=9

```

**Summary 1: Flash increase by 676 bytes, RAM reduced by 448 bytes.**


Increasing the chain by 1 (to 10):

```
[181/182] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       68736 B         1 MB      6.56%
             RAM:       28940 B       256 KB     11.04%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/jack/workspace/zephyrproject/zephyr/build/observer_nrf52840/zephyr/zephyr.elf for board: nrf52840dk
(python) [jack@vich-archlinux-dell-laptop observer_nrf52840]$ git diff
diff --git a/samples/bluetooth/observer/prj_extended.conf b/samples/bluetooth/observer/prj_extended.conf
index cab86c38d14..07439181f19 100644
--- a/samples/bluetooth/observer/prj_extended.conf
+++ b/samples/bluetooth/observer/prj_extended.conf
@@ -14,4 +14,9 @@ CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX=1650
 
 # Increase Zephyr Bluetooth LE Controller Rx buffer to receive complete chain
 # of PDUs
-CONFIG_BT_CTLR_RX_BUFFERS=9
+CONFIG_BT_CTLR_RX_BUFFERS=18
+
+CONFIG_BT_CTLR_ADVANCED_FEATURES=y
+CONFIG_BT_CTLR_SCAN_AUX_SET=10
+# CONFIG_BT_CTLR_SCAN_AUX_USE_CHAINS=y
+# CONFIG_BT_CTLR_SCAN_AUX_CHAIN_COUNT=10

```

```
[181/182] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       69412 B         1 MB      6.62%
             RAM:       28428 B       256 KB     10.84%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/jack/workspace/zephyrproject/zephyr/build/observer_nrf52840/zephyr/zephyr.elf for board: nrf52840dk
(python) [jack@vich-archlinux-dell-laptop observer_nrf52840]$ git diff
diff --git a/samples/bluetooth/observer/prj_extended.conf b/samples/bluetooth/observer/prj_extended.conf
index cab86c38d14..2c997cf3b77 100644
--- a/samples/bluetooth/observer/prj_extended.conf
+++ b/samples/bluetooth/observer/prj_extended.conf
@@ -14,4 +14,9 @@ CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX=1650
 
 # Increase Zephyr Bluetooth LE Controller Rx buffer to receive complete chain
 # of PDUs
-CONFIG_BT_CTLR_RX_BUFFERS=9
+CONFIG_BT_CTLR_RX_BUFFERS=18
+
+CONFIG_BT_CTLR_ADVANCED_FEATURES=y
+# CONFIG_BT_CTLR_SCAN_AUX_SET=9
+CONFIG_BT_CTLR_SCAN_AUX_USE_CHAINS=y
+CONFIG_BT_CTLR_SCAN_AUX_CHAIN_COUNT=10

```

**Summary 2: Flash increase 0 bytes, RAM increased by 64 bytes (compared to 128 bytes for BT_SCAN_AUX_SET use).**

**Review by cvinayak — APPROVED** [review]
I get the visual impression of running the observer that BT_CTLR_SCAN_AUX_SET generate more advertising report compared to BT_CTLR_SCAN_AUX_CHAINS.

May be @jakkra could help us here on the rate of advertising reports in his ecosystem.

**Comment by Tronil** [comment]
> **Summary 2: Flash increase 0 bytes, RAM increased by 64 bytes (compared to 128 bytes for BT_SCAN_AUX_SET use).**

Interesting - downstream we only get an increase per count of 44 bytes. Maybe due to padding differences?... I don't see anything that should affect the size configuration wise.

**Comment by cvinayak** [comment]
> > **Summary 2: Flash increase 0 bytes, RAM increased by 64 bytes (compared to 128 bytes for BT_SCAN_AUX_SET use).**
> 
> Interesting - downstream we only get an increase per count of 44 bytes. Maybe due to padding differences?... I don't see anything that should affect the size configuration wise.
> 
> > I get the visual impression of running the observer that BT_CTLR_SCAN_AUX_SET generate more advertising report compared to BT_CTLR_SCAN_AUX_CHAINS.
> 
> One thing we may want to look at is the value of `ticks_min_delta` in `chain_insert_in_sched_list` - i.e. the minimum difference in timestamp required for insertion into the scheduling list. I have currently set it to EVENT_OVERHEAD_XTAL_US + EVENT_OVERHEAD_START_US, but that may be too conservative. I am also not really sure what EVENT_OVERHEAD_XTAL_US is supposed to cover.

Ah, use of `EVENT_OVERHEAD_XTAL_US`  can explain, you do not need this value only `EVENT_OVERHEAD_END_US + EVENT_OVERHEAD_START_US` should suffice as the CPU overhead between radio end (last bit on air) to radio start (first bit on air), in thoery. I did not consider here the radio ramp up etc... though.

File: subsys/bluetooth/controller/ll_sw/ull_scan_aux.c [hunk 1]
```diff
@@ -1470,3 +1498,1412 @@ static void ticker_op_cb(uint32_t status, void *param)
...
+
+	/* Find insertion point */
+	while (curr && chain_ticker_ticks_diff(chain->ticker_ticks, curr->ticker_ticks) > 0) {
+		prev = curr;
+		curr = curr->next;
+	}
+
+	/* Check for conflict with existing entry */
+	ticks_min_delta = HAL_TICKER_US_TO_TICKS(EVENT_OVERHEAD_XTAL_US) +
+			  HAL_TICKER_US_TO_TICKS(EVENT_OVERHEAD_START_US);
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:2841]:
```suggestion
	ticks_min_delta = HAL_TICKER_US_TO_TICKS(EVENT_OVERHEAD_END_US + EVENT_OVERHEAD_START_US);
```

**Review by cvinayak — CHANGES_REQUESTED** [review]

**Comment by jakkra** [comment]
> I get the visual impression of running the observer that BT_CTLR_SCAN_AUX_SET generate more advertising report compared to BT_CTLR_SCAN_AUX_CHAINS.
> 
> May be @jakkra could help us here on the rate of advertising reports in his ecosystem.
```
CONFIG_BT_CTLR_RX_BUFFERS=18
CONFIG_BT_CTLR_SCAN_AUX_SET=9
```
=> ~200 packets/s

```
CONFIG_BT_CTLR_RX_BUFFERS=18
CONFIG_BT_CTLR_SCAN_AUX_USE_CHAINS=y
CONFIG_BT_CTLR_SCAN_AUX_CHAIN_COUNT=9
```
=> ~200 packets/s

Note: Did not change the `ticks_min_delta` suggestion.

**Review by carlescufi — DISMISSED** [review]

**Comment by Tronil** [comment]
> @Tronil One minor change request that I requested, then we are good to get this merged :-)

Done :+1:

**Review by cvinayak — DISMISSED** [review]

**Review by cvinayak — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**